### DISCUSSION TITLE
Schrödinger's session in WebTestCase

### DISCUSSION THREAD

**Discussion by Trismegiste** [symfony/symfony#59237] [Q&A] [answered]
Hello,

There is something I obviously don't understand about sessions in functional tests with WebTestCase and KernelBrowser.
* I have a controller that set and get data in session through the service "SomeService"
* I have a service "SomeService" that requires the RequestStack in the constructor
* and the test class where the last line ```$this->assertEquals('yolo', $service->getInfo())``` fails with an exception "There is currently no session available." 

Though the previous test confirms : 
* there is a session
* the property 'info' in session is initialized to 'yolo'

But when I get the service directly, there is no session o_0
Current environment : Docker with PHP 8.3 + Symfony 7.1 + FrankenPHP

```bash
1) App\Tests\Controller\YoloControllerTest::testBug
Symfony\Component\HttpFoundation\Exception\SessionNotFoundException: There is currently no session available.

/app/vendor/symfony/http-foundation/RequestStack.php:105
/app/src/Service/SomeService.php:26
/app/tests/Controller/YoloControllerTest.php:35
/app/vendor/phpunit/phpunit/phpunit:107
```
## The controller YoloController
```php
namespace App\Controller;

use App\Service\SomeService;
use Symfony\Bundle\FrameworkBundle\Controller\AbstractController;
use Symfony\Component\HttpFoundation\Response;
use Symfony\Component\Routing\Attribute\Route;

class YoloController extends AbstractController
{

    public function __construct(protected SomeService $service)    {    }

    #[Route('/set-info')]
    public function setInfo(): Response
    {
        $this->service->setInfo('yolo');
        return new Response('set');
    }

    #[Route('/get-info')]
    public function getInfo(): Response
    {
        return new Response($this->service->getInfo());
    }
}
``` 
## The service SomeService
```php
namespace App\Service;

use Symfony\Component\HttpFoundation\RequestStack;

class SomeService
{

    public function __construct(protected RequestStack $stack)    {    }

    public function setInfo(string $value): void
    {
        $this->stack->getSession()->set('info', $value);
    }

    public function getInfo(): string
    {
        return $this->stack->getSession()->get('info');
    }
}
```

## The functional test class
```php
namespace App\Tests\Controller;

use App\Service\SomeService;
use Override;
use Symfony\Bundle\FrameworkBundle\KernelBrowser;
use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;

class YoloControllerTest extends WebTestCase
{

    protected KernelBrowser $client;

    #[Override]
    protected function setUp(): void
    {
        $this->client = static::createClient();
    }

    public function testBug()
    {
        $this->client->request('GET', '/set-info');
        $this->assertResponseIsSuccessful();

        $this->client->request('GET', '/get-info');
        $this->assertResponseIsSuccessful();
        $this->assertEquals('yolo', $this->client->getResponse()->getContent());  // the response contains the string 'yolo', it's ok

        $service = $this->client->getContainer()->get(SomeService::class);
        $this->assertEquals('yolo', $service->getInfo());  // throws an exception "No session"
    }
}
```

Anyone can help ? Thank you in advance

**Comment by alexandre-le-borgne** [comment, marked as answer]
See https://symfony.com/doc/current/testing.html#multiple-requests-in-one-test

> After making a request, subsequent requests will make the client reboot the kernel. This recreates the container from scratch to ensures that requests are isolated and use new service objects each time. This behavior can have some unexpected consequences: for example, the security token will be cleared, Doctrine entities will be detached, etc.

This may be the same issue (with workaround) :

https://github.com/symfony/symfony/discussions/45662#discussioncomment-4352880

**Reply by Trismegiste** [reply]
Thank you very much for your reply

> See https://symfony.com/doc/current/testing.html#multiple-requests-in-one-test

My bad I miss this point in the doc.

> [#45662 (comment)](https://github.com/symfony/symfony/discussions/45662#discussioncomment-4352880)

Could be interesting in some cases but not mine, I won't build manually a ton of services solely for the tests.

**Comment by stof** [comment]
Throwing an exception in `$service->getInfo()` in your test is the expected behavior. At this point, the RequestStack has no active request (as you do that _outside_ the request handling) and so it cannot read the session associated with the current request.

the fact that the session is related to the current request is precisely why the Session is not available as a service anymore in Symfony (it used to be in older versions, but this caused lots of weird issues, especially when using a long-running process handling multiple request).

**Reply by Trismegiste** [reply]
Thank you for your reply

> Throwing an exception in `$service->getInfo()` in your test is the expected behavior. At this point, the RequestStack has no active request (as you do that _outside_ the request handling) and so it cannot read the session associated with the current request.

I understand the problem and it seems pretty secure for good reasons.

> the fact that the session is related to the current request is precisely why the Session is not available as a service anymore in Symfony (it used to be in older versions

Yeah, that's why I didn't expect this kind a wall. 
I'll come up with something for ditching the RequestStack in my services, it was a terrible idea anyway (what about CLI for example)

**Reply by stof** [reply]
if you need to make your service usable in the CLI, it cannot expect a Session to be available all the time (there is no current session in the CLI).
